### PR TITLE
Add formatter and linter in the CI.

### PR DESCRIPTION
**Pull request by yaoyuannnn**
This adds the following in the CI:
1) A code formatter using `black`.
2) A code linter using `flake8`. It works as a complement to `black` by
   checking syntax errors, possible bugs, stylistic errors, etc.

All the Python files are fixed to be compliant with the formatter and the
linter.

### CODE REVIEW

File: .circleci/run_style_check.py
Comment (dancysoft) [inline .circleci/run_style_check.py]: `with os.popen(... ) as stream:`

File: .circleci/run_style_check.py
Comment (dancysoft) [inline .circleci/run_style_check.py]:
```suggestion
        for y in (x.strip() for x in stream.readlines())
```

File: .circleci/run_style_check.py
Comment (dancysoft) [reply .circleci/run_style_check.py]: That said, you might consider using `subprocess.check_output("git diff-tree --no-commit-id --name-only --diff-filter=ACM -r HEAD", shell=True, text=True).splitlines()`.

File: .circleci/run_style_check.py
Comment (dancysoft) [reply .circleci/run_style_check.py]: Also, did you really find that the `strip()` was necessary?  If `git diff-tree --name-only` is documented to print one filename per line, there should be need for further manipulation beyond the splitlines.

File: .circleci/run_style_check.py
Comment (dancysoft) [inline .circleci/run_style_check.py]: Can we use `repo.working_dir` or `repo.working_tree_dir` for this?  https://gitpython.readthedocs.io/en/stable/tutorial.html#meet-the-repo-type mentions `working_tree_dir` several times.

File: .circleci/run_style_check.py
Comment (dancysoft) [inline .circleci/run_style_check.py]: Instead of repeatedly invoking `black`, call it once with the list of filenames.

File: .circleci/run_style_check.py
Comment (dancysoft) [inline .circleci/run_style_check.py]:
```suggestion
        subprocess.run(command, text=True, check=True)
```

File: .circleci/run_style_check.py
Comment (yaoyuannnn) [reply .circleci/run_style_check.py]: Yes, thanks! This simplifies it quite a bit :)

File: .circleci/run_style_check.py
Comment (yaoyuannnn) [reply .circleci/run_style_check.py]: `strip()` should be unnecessary here. Removed it.

File: .circleci/run_style_check.py
Comment (yaoyuannnn) [reply .circleci/run_style_check.py]: Yes, thanks! Done.

File: .circleci/run_style_check.py
Comment (yaoyuannnn) [reply .circleci/run_style_check.py]: Yes! Done.

File: .circleci/run_style_check.py
Comment (yaoyuannnn) [reply .circleci/run_style_check.py]: Awesome! I added `--check` which will return non-zero value if any file needs reformat.

**Comment by dancysoft** [comment]
Proposal: 
* Run `black $(git ls-files '*.py')` on this repo, commit, and submit an MR which brings all source files into compliance.
* Change this commit to always run `black $(git ls-files '*.py')`.

**Comment by yaoyuannnn** [comment]
> Proposal:
> 
> * Run `black $(git ls-files '*.py')` on this repo, commit, and submit an MR which brings all source files into compliance.
> * Change this commit to always run `black $(git ls-files '*.py')`.

Sounds good! All the Python files are reformatted in this MR.

File: .circleci/run_style_check.py
Comment (dancysoft) [inline .circleci/run_style_check.py]: all *.py files

File: .circleci/run_style_check.py
Comment (dancysoft) [inline .circleci/run_style_check.py]:
```suggestion
    repo = Repo(os.path.realpath(__file__), search_parent_directories=True)
    repo_path = repo.working_tree_dir

    command = "black"
    if not apply_diff:
        command += " --check --diff --color"

    command += ' $(git ls-files "*.py")'

    subprocess.run(command, cwd=repo_path, shell=True, text=True, check=True)
```

**Review by dancysoft — COMMENTED** [review]
btw, did you consider using `flake8`? It has formatting checks as well as checks for things like unused imports (which there are a few of).

File: .circleci/run_style_check.py
Comment (yaoyuannnn) [reply .circleci/run_style_check.py]: Thanks! Done!

File: .circleci/run_style_check.py
Comment (yaoyuannnn) [reply .circleci/run_style_check.py]: Done.

**Comment by yaoyuannnn** [comment]
> btw, did you consider using `flake8`? It has formatting checks as well as checks for things like unused imports (which there are a few of).

I used `flake8` before for removing unused imports. `black` was used because it's apparently more popular on vscode. I have no bias on either and am open to a switch to `flake8` or maybe employing both.

**Comment by dancysoft** [comment]
> > btw, did you consider using `flake8`? It has formatting checks as well as checks for things like unused imports (which there are a few of).
> 
> I used `flake8` before for removing unused imports. `black` was used because it's apparently more popular on vscode. I have no bias on either and am open to a switch to `flake8` or maybe employing both.

In that case I'd like to advocate for using `flake8` instead.  Looks like running `flake8 --max-line-length 88` passes except for unused imports warnings.    You don't even have to specify the list of files to check which is nice.

**Comment by yaoyuannnn** [comment]
> > > btw, did you consider using `flake8`? It has formatting checks as well as checks for things like unused imports (which there are a few of).
> > 
> > 
> > I used `flake8` before for removing unused imports. `black` was used because it's apparently more popular on vscode. I have no bias on either and am open to a switch to `flake8` or maybe employing both.
> 
> In that case I'd like to advocate for using `flake8` instead. Looks like running `flake8 --max-line-length 88` passes except for unused imports warnings. You don't even have to specify the list of files to check which is nice.

Sounds good to me. Seems like a linter like `flake8` works best together with a reformatter such as `black`. I added `flake8` in the CI too as a complement to `black`. All the files now have their style fixed to be compliant with `flake8`.

**Comment by dancysoft** [comment]
This looks good to me.  Can we wait until #41 is approved/merged before finalizing this?  Otherwise there will be painful conflicts for @thcipriani to resolve.

**Comment by yaoyuannnn** [comment]
> This looks good to me. Can we wait until #41 is approved/merged before finalizing this? Otherwise there will be painful conflicts for @thcipriani to resolve.

Yes, for sure.

**Review by dancysoft — APPROVED** [review]
Thanks!

**Pull request merged**
